contracts/src/OffChainMessageContracts/ValidatorSetSig.sol

```diff
@@ -0,0 +1,130 @@
+ * targetContractAddress: Address of the contract that the payload should be forwarded to
+ * targetBlockchainID: Blockchain ID of the chain the message is intended for
+ * nonce: Unique nonce for the target contract address to provide replay protection
+ * payload: Payload to be forwarded to the target contract. Usually ABI encoded function call with parameters.
+ */
+struct ValidatorSetSigMessage {
+    address validatorSetSigAddress;
+    address targetContractAddress;
+    bytes32 targetBlockchainID;
+    uint256 nonce;
+    bytes payload;
+}
+
+/**
+ * @dev Contract that verifies that a set threshold of validators from a given blockchainID
+ * have signed an off-chain Warp message and forwards the payload to the target contract specified in the payload.
+ * The threshold itself is set by the validator themselves in their Warp configs:
+ * https://github.com/ava-labs/subnet-evm/blob/master/precompile/contracts/warp/config.go#L50
+ *
+ * This is intended to be used for safe off-chain governance of enabled contracts. An example usecase would be
+ * to deploy a `Ownable` target contract that is owned by an instance of this contract and adding the
+ * `onlyOwner` modifier to the functions that should be governed.
+ */
+contract ValidatorSetSig is ReentrancyGuard {
+    /**
+     * @notice A blockchain ID whose validators need to sign the message for it to be considered valid.
+     */
+    bytes32 public immutable validatorBlockchainID;
+
+    /**
+     * @dev Tracks nonces for messages sent to a specific contract address to provide replay protection.
+     * The target contract address is used as the key instead of keeping a global nonce to prevent
+     * one party from consuming a global nonce that another was intending to use
+     * in case a single instance of ValidatorSetSig contract is used to manage multiple downstream contracts.
+     */
+    mapping(address targetContractAddress => uint256 nonce) public nonces;
+
+    /**
+     * @notice Address that the off-chain Warp message sets as the "source" address.
+     * @dev The address is not owned by any EOA or smart contract account, so it
+     * cannot possibly be the source address of any other Warp message emitted by the VM.
+     */
+    address public constant VALIDATORS_SOURCE_ADDRESS = address(0);
+
+    /**
+     * @notice The blockchain ID of the chain the contract is deployed on.
+     */
+    bytes32 public immutable blockchainID;
+
+    /**
+     * @notice Warp precompile used for sending and receiving Warp messages.
+     */
+    IWarpMessenger public constant WARP_MESSENGER =
+        IWarpMessenger(0x0200000000000000000000000000000000000005);
+
+    /**
+     * @notice Emited when the payload is successfully delivered to the target contract.
+     */
+    event Delivered(address indexed targetContractAddress, uint256 indexed nonce);
+
+    constructor(bytes32 validatorBlockChainID) {
+        validatorBlockchainID = validatorBlockChainID;
```

Comment:
nit: I'm not sure what our coding style guide says about this, but having these two variable names differ only by their capitalization for "chain" seems a bit error prone.

Reply:
we'd make the parameter `validatorBlockchainID_`